File: livingrimoire start here/LivinGrimoire C_sharp/LivinGrimoire.cs
using System;
using System.Collections.Generic;
using System.Collections;
public class AbsDictionaryDB
{
    public virtual void Save(string key, string value)
    {
        // Save to DB (override me)
    }

    public virtual string Load(string key)
    {
        // Override me
        return "null";
    }
}
public abstract class Mutatable
{
    public bool algKillSwitch = false;

    public abstract string Action(string ear, string skin, string eye);
    public abstract bool Completed();

    public string MyName()
    {
        // Returns the class name
        return GetType().Name;
    }
}
public class APSay : Mutatable
{
    // It speaks something x times
    // A most basic skill.
    // Also fun to make the chobit say what you want.

    protected string param;
    private int at;

    public APSay(int repetitions, string param)
    {
        if (repetitions > 10)
        {
            repetitions = 10;
        }
        this.at = repetitions;
        this.param = param;
    }

    public override string Action(string ear, string skin, string eye)
    {
        // TODO: Implement your logic here
        string axnStr = "";
        if (this.at > 0)
        {
            if (!string.Equals(ear, param, StringComparison.OrdinalIgnoreCase))
            {
                axnStr = param;
                this.at--;
            }
        }
        return axnStr;
    }

    public override bool Completed()
    {
        // TODO: Implement your logic here
        return this.at < 1;
    }

}
public class APVerbatim : Mutatable
{
    private List<string> sentences = new List<string>();
    private int at = 0;

    public APVerbatim(params string[] sentences)
    {
        foreach (string sentence in sentences)
        {
            this.sentences.Add(sentence);
        }

        if (sentences.Length == 0)
        {
            at = 30;
        }
    }

    public APVerbatim(List<string> list1)
    {
        this.sentences = new List<string>(list1);

        if (this.sentences.Count == 0)
        {
            at = 30;
        }
    }

    public override string Action(string ear, string skin, string eye)
    {
        string axnStr = "";
        if (at < sentences.Count)
        {
            axnStr = sentences[at];
            at++;
        }
        return axnStr;
    }

    public override bool Completed()
    {
        return at >= sentences.Count;
    }
}
public class GrimoireMemento
{
    private AbsDictionaryDB absDictionaryDB;

    public GrimoireMemento(AbsDictionaryDB absDictionaryDB)
    {
        this.absDictionaryDB = absDictionaryDB;
    }

    public string SimpleLoad(string key)
    {
        return this.absDictionaryDB.Load(key);
    }

    public void SimpleSave(string key, string value)
    {
        if (key == "" || value == "")
        {
            return;
        }
        this.absDictionaryDB.Save(key, value);
    }
}
public class Algorithm
{
    private List<Mutatable> algParts = new List<Mutatable>();

    public Algorithm(List<Mutatable> algParts)
    {
        this.algParts = algParts;
    }

    public List<Mutatable> GetAlgParts()
    {
        return algParts;
    }

    public int GetSize()
    {
        return algParts.Count;
    }
}
public class Kokoro
{
    private string emot = "";

    public string GetEmot()
    {
        return emot;
    }

    public void SetEmot(string emot)
    {
        this.emot = emot;
    }

    public GrimoireMemento grimoireMemento;
    public Hashtable toHeart = new Hashtable();

    public Kokoro(AbsDictionaryDB absDictionaryDB)
    {
        this.grimoireMemento = new GrimoireMemento(absDictionaryDB);
    }
}
public class Neuron
{
    private Dictionary<int, List<Algorithm>> defcons = new Dictionary<int, List<Algorithm>>();

    public Neuron()
    {
        for (int i = 1; i <= 5; i++)
        {
            defcons[i] = new List<Algorithm>();
        }
    }

    public void InsertAlg(int priority, Algorithm alg)
    {
        if (0 < priority && priority < 6)
        {
            if (defcons[priority].Count < 4)
            {
                defcons[priority].Add(alg);
            }
        }
    }

    public Algorithm? GetAlg(int defcon)
    {
        if (defcons[defcon].Count > 0)
        {
            Algorithm temp = defcons[defcon][0];
            defcons[defcon].RemoveAt(0);
            return temp;
        }
        return null;
    }
}
public class Skill
{
    protected Kokoro? kokoro = null; // consciousness, shallow ref class to enable interskill communications
    protected Algorithm? outAlg = null; // skills output
    protected int outpAlgPriority = -1; // defcon 1->5

    public Skill()
    {
    }

    public virtual void Input(string ear, string skin, string eye)
    {
    }

    public virtual void Output(Neuron noiron)
    {
        if (outAlg != null)
        {
            noiron.InsertAlg(outpAlgPriority, outAlg);
            outpAlgPriority = -1;
            outAlg = null;
        }
    }

    public virtual void SetKokoro(Kokoro kokoro)
    {
        this.kokoro = kokoro;
    }

    protected void SetVerbatimAlg(int priority, params string[] sayThis)
    {
        outAlg = SimpleVerbatimAlgorithm(sayThis);
        outpAlgPriority = priority; // 1->5 1 is the highest algorithm priority
    }

    protected void SetSimpleAlg(params string[] sayThis)
    {
        outAlg = SimpleVerbatimAlgorithm(sayThis);
        outpAlgPriority = 4; // 1->5 1 is the highest algorithm priority
    }

    protected void SetVerbatimAlgFromList(int priority, List<string> sayThis)
    {
        outAlg = AlgBuilder(new APVerbatim(sayThis));
        outpAlgPriority = priority; // 1->5 1 is the highest algorithm priority
    }

    protected void AlgPartsFusion(int priority, params Mutatable[] algParts)
    {
        outAlg = AlgBuilder(algParts);
        outpAlgPriority = priority; // 1->5 1 is the highest algorithm priority
    }

    public bool PendingAlgorithm()
    {
        return outAlg != null;
    }
    // skill utils (alg building methods)
    // Alg part based algorithm building methods (var args param)
    public Algorithm AlgBuilder(params Mutatable[] algParts)
    {
        // Returns an algorithm built with the algPart varargs
        List<Mutatable> algParts1 = new List<Mutatable>();
        foreach (Mutatable part in algParts)
        {
            algParts1.Add(part);
        }
        Algorithm algorithm = new Algorithm(algParts1);
        return algorithm;
    }

    // String based algorithm building methods
    public Algorithm SimpleVerbatimAlgorithm(params string[] sayThis)
    {
        // Returns an algorithm that says the sayThis Strings verbatim per think cycle
        return AlgBuilder(new APVerbatim(sayThis));
    }

    public string StrContainsList(string str1, List<string> items)
    {
        // Returns the 1st match between words in a string and values in a list.
        foreach (string temp in items)
        {
            if (str1.Contains(temp))
            {
                return temp;
            }
        }
        return "";
    }
    public virtual string SkillNotes(string param)
    {
        return "notes unknown";
    }

}
public class DiHelloWorld : Skill
{
    // hello world skill for testing purposes
    public DiHelloWorld() : base()
    {
    }

    public override void Input(string ear, string skin, string eye)
    {
        switch (ear)
        {
            case "hello":
                base.SetVerbatimAlg(4, "hello world"); // 1->5 1 is the highest algorithm priority
                break;
        }
    }
    public override string SkillNotes(string param)
    {
        if (param == "notes")
        {
            return "plain hello world skill";
        }
        else if (param == "triggers")
        {
            return "say hello";
        }
        return "note unavailable";
    }

}
public class Cerabellum
{
    // Runs an algorithm
    private int fin;
    private int at;
    private bool incrementAt = false;
    public Algorithm? alg;
    private bool ia = false; // isActive attribute
    private string emot = "";

    public void AdvanceInAlg()
    {
        if (incrementAt)
        {
            incrementAt = false;
            at++;
            if (at == fin)
            {
                ia = false;
            }
        }
    }

    public int GetAt()
    {
        return at;
    }

    public string GetEmot()
    {
        return emot;
    }

    public bool SetAlgorithm(Algorithm algorithm)
    {
        if (!IsActive() && algorithm.GetAlgParts().Count != 0)
        {
            alg = algorithm;
            at = 0;
            fin = algorithm.GetSize();
            ia = true;
            emot = alg.GetAlgParts()[at].MyName(); // Updated line
            return false;
        }
        return true;
    }

    public bool IsActive()
    {
        return ia;
    }

    public string Act(string ear, string skin, string eye)
    {
        string axnStr = "";
        if (!IsActive())
        {
            return axnStr;
        }
        if (at < fin)
        {
            axnStr = alg!.GetAlgParts()[at].Action(ear, skin, eye);
            emot = alg!.GetAlgParts()[at].MyName();
            if (alg!.GetAlgParts()[at].Completed())
            {
                incrementAt = true;
            }
        }
        return axnStr;
    }

    public void DeActivation()
    {
        ia = IsActive() && !alg!.GetAlgParts()[at].algKillSwitch;
    }
}
public class Fusion
{
    private string emot = "";
    private string result = "";
    private Cerabellum[] ceraArr = new Cerabellum[5];

    public Fusion()
    {
        for (int i = 0; i < 5; i++)
        {
            ceraArr[i] = new Cerabellum();
        }
    }

    public string GetEmot()
    {
        return emot;
    }

    public void LoadAlgs(Neuron neuron)
    {
        for (int i = 1; i <= 5; i++)
        {
            if (!ceraArr[i - 1].IsActive())
            {
                Algorithm temp = neuron.GetAlg(i)!;
                if (temp != null)
                {
                    ceraArr[i - 1].SetAlgorithm(temp);
                }
            }
        }
    }

    public string RunAlgs(string ear, string skin, string eye)
    {
        result = "";
        for (int i = 0; i < 5; i++)
        {
            if (!ceraArr[i].IsActive())
            {
                continue;
            }
            result = ceraArr[i].Act(ear, skin, eye);
            ceraArr[i].AdvanceInAlg();
            emot = ceraArr[i].GetEmot();
            ceraArr[i].DeActivation(); // Deactivation if Mutatable.algkillswitch = true
            return result;
        }
        emot = "";
        return result;
    }
}
public class Chobits
{
    public List<Skill> dClasses = new List<Skill>();
    protected Fusion fusion;
    protected Neuron noiron;
    protected Kokoro kokoro = new Kokoro(new AbsDictionaryDB()); // consciousness
    private bool isThinking = false;
    private readonly List<Skill> awareSkills = new List<Skill>();

    public Chobits()
    {
        // c'tor
        this.fusion = new Fusion();
        this.noiron = new Neuron();
    }

    public void SetDataBase(AbsDictionaryDB absDictionaryDB)
    {
        this.kokoro = new Kokoro(absDictionaryDB);
    }

    public Chobits AddSkill(Skill skill)
    {
        // add a skill (builder design patterned func)
        if (this.isThinking)
        {
            return this;
        }
        skill.SetKokoro(this.kokoro);
        this.dClasses.Add(skill);
        return this;
    }

    public Chobits AddSkillAware(Skill skill)
    {
        // add a skill with Chobit Object in their constructor
        skill.SetKokoro(this.kokoro);
        this.awareSkills.Add(skill);
        return this;
    }

    public void ClearSkills()
    {
        // remove all skills
        if (this.isThinking)
        {
            return;
        }
        this.dClasses.Clear();
    }

    public void AddSkills(params Skill[] skills)
    {
        if (this.isThinking)
        {
            return;
        }
        foreach (Skill skill in skills)
        {
            skill.SetKokoro(this.kokoro);
            this.dClasses.Add(skill);
        }
    }

    public void RemoveSkill(Skill skill)
    {
        if (this.isThinking)
        {
            return;
        }
        this.dClasses.Remove(skill);
    }

    public bool ContainsSkill(Skill skill)
    {
        return this.dClasses.Contains(skill);
    }

    public string Think(string ear, string skin, string eye)
    {
        this.isThinking = true;
        foreach (Skill dCls in this.dClasses)
        {
            InOut(dCls, ear, skin, eye);
        }
        this.isThinking = false;
        foreach (Skill dCls2 in this.awareSkills)
        {
            InOut(dCls2, ear, skin, eye);
        }
        this.fusion.LoadAlgs(this.noiron);
        return this.fusion.RunAlgs(ear, skin, eye);
    }

    public string GetSoulEmotion()
    {
        // get the last active AlgPart name
        // the AP is an action, and it also represents
        // an emotion
        return this.fusion.GetEmot();
    }

    protected void InOut(Skill dClass, string ear, string skin, string eye)
    {
        dClass.Input(ear, skin, eye); // new
        dClass.Output(this.noiron);
    }

    public Kokoro GetKokoro()
    {
        // several chobits can use the same soul
        // this enables telepathic communications
        // between chobits in the same project
        return this.kokoro;
    }

    public void SetKokoro(Kokoro kokoro)
    {
        // use this for telepathic communication between different chobits objects
        this.kokoro = kokoro;
    }

    public Fusion GetFusion()
    {
        return this.fusion;
    }

    public List<string> GetSkillList()
    {
        List<string> result = new List<string>();
        foreach (Skill skill in this.dClasses)
        {
            result.Add(skill.GetType().Name);
        }
        return result;
    }
}
public class Brain
{
    public Chobits logicChobit;
    public Chobits hardwareChobit;
    private string emotion = "";
    private string bodyInfo = "";
    private string logicChobitOutput = "";

    public string GetEmotion => emotion;

    public string GetBodyInfo => bodyInfo;

    public string GetLogicChobitOutput => logicChobitOutput;

    public Brain()
    {
        logicChobit = new Chobits();
        hardwareChobit = new Chobits();
        hardwareChobit.SetKokoro(logicChobit.GetKokoro());
    }

    public void DoIt(string ear, string skin, string eye)
    {
        if (!string.IsNullOrEmpty(bodyInfo))
        {
            logicChobitOutput = logicChobit.Think(ear, bodyInfo, eye);
        }
        else
        {
            logicChobitOutput = logicChobit.Think(ear, skin, eye);
        }
        emotion = logicChobit.GetSoulEmotion();
        // Case: Hardware skill wishes to pass info to logical chobit
        bodyInfo = hardwareChobit.Think(logicChobitOutput, skin, eye);
    }
    public void AddLogicalSkill(Skill skill)
    {
        logicChobit.AddSkill(skill);
    }

    public void AddHardwareSkill(Skill skill)
    {
        hardwareChobit.AddSkill(skill);
    }

}
public class DiPrinter : Skill
{
    // hello world skill for testing purposes
    public DiPrinter() : base()
    {
    }

    public override void Input(string ear, string skin, string eye)
    {
        if (ear == "")
        {
            return;
        }
        Console.WriteLine(ear);
    }
}

File: livingrimoire start here/LivinGrimoire JavaScript/LivinGrimoire.js
// livingrimoire.js

class AbsDictionaryDB {
    save(key, value) {
        // Save to DB (override me)
    }

    load(key) {
        // Override me
        return "null";
    }
}

class Mutatable {
    constructor() {
        this.algKillSwitch = false;
    }

    // Abstract methods (to be overridden by subclasses)
    Action(ear, skin, eye) {
        throw new Error('Action method must be overridden');
    }

    Completed() {
        throw new Error('Completed method must be overridden');
    }

    MyName() {
        // Returns the class name
        return this.constructor.name;
    }
}

class APSay extends Mutatable {
    constructor(repetitions, param) {
        super();
        this.param = param;
        this.at = repetitions > 10 ? 10 : repetitions;
    }

    Action(ear, skin, eye) {
        let axnStr = "";
        if (this.at > 0) {
            if (ear.toLowerCase() !== this.param.toLowerCase()) {
                axnStr = this.param;
                this.at--;
            }
        }
        return axnStr;
    }

    Completed() {
        return this.at < 1;
    }
}

class APVerbatim extends Mutatable {
    constructor(...sentences) {
        super();
        this.sentences = sentences.length > 0 ? sentences : [];
        this.at = sentences.length === 0 ? 30 : 0;
    }

    Action(ear, skin, eye) {
        let axnStr = "";
        if (this.at < this.sentences.length) {
            axnStr = this.sentences[this.at];
            this.at++;
        }
        return axnStr;
    }

    Completed() {
        return this.at >= this.sentences.length;
    }
}

class GrimoireMemento {
    constructor(absDictionaryDB) {
        this.absDictionaryDB = absDictionaryDB;
    }

    SimpleLoad(key) {
        return this.absDictionaryDB.load(key);
    }

    SimpleSave(key, value) {
        if (!key || !value) {
            return;
        }
        this.absDictionaryDB.save(key, value);
    }
}

class Algorithm {
    constructor(algParts) {
        this.algParts = algParts;
    }

    GetAlgParts() {
        return this.algParts;
    }

    GetSize() {
        return this.algParts.length;
    }
}
class Kokoro {
    constructor(absDictionaryDB) {
        this.emot = "";
        this.grimoireMemento = new GrimoireMemento(absDictionaryDB);
        this.toHeart = new Map();
    }

    GetEmot() {
        return this.emot;
    }

    SetEmot(emot) {
        this.emot = emot;
    }
}

class Neuron {
    constructor() {
        this.defcons = new Map();
        for (let i = 1; i <= 5; i++) {
            this.defcons.set(i, []);
        }
    }

    InsertAlg(priority, alg) {
        if (priority > 0 && priority < 6) {
            const algList = this.defcons.get(priority);
            if (algList.length < 4) {
                algList.push(alg);
            }
        }
    }

    GetAlg(defcon) {
        const algList = this.defcons.get(defcon);
        if (algList && algList.length > 0) {
            return algList.shift();
        }
        return null;
    }
}

class Skill {
    constructor() {
        this.kokoro = null; // consciousness, shallow ref class to enable interskill communications
        this.outAlg = null; // skills output
        this.outpAlgPriority = -1; // defcon 1->5
    }

    Input(ear, skin, eye) {
        // Virtual method to be overridden by subclasses
    }

    Output(noiron) {
        if (this.outAlg !== null) {
            noiron.InsertAlg(this.outpAlgPriority, this.outAlg);
            this.outpAlgPriority = -1;
            this.outAlg = null;
        }
    }

    SetKokoro(kokoro) {
        this.kokoro = kokoro;
    }

    SetVerbatimAlg(priority, ...sayThis) {
        this.outAlg = this.SimpleVerbatimAlgorithm(...sayThis);
        this.outpAlgPriority = priority; // 1->5 1 is the highest algorithm priority
    }

    SetSimpleAlg(...sayThis) {
        this.outAlg = this.SimpleVerbatimAlgorithm(...sayThis);
        this.outpAlgPriority = 4; // 1->5 1 is the highest algorithm priority
    }

    SetVerbatimAlgFromList(priority, sayThis) {
        this.outAlg = this.AlgBuilder(new APVerbatim(sayThis));
        this.outpAlgPriority = priority; // 1->5 1 is the highest algorithm priority
    }

    AlgPartsFusion(priority, ...algParts) {
        this.outAlg = this.AlgBuilder(...algParts);
        this.outpAlgPriority = priority; // 1->5 1 is the highest algorithm priority
    }

    PendingAlgorithm() {
        return this.outAlg !== null;
    }

    // skill utils (alg building methods)
    // Alg part based algorithm building methods (var args param)
    AlgBuilder(...algParts) {
        // Returns an algorithm built with the algPart varargs
        const algParts1 = [];
        for (const part of algParts) {
            algParts1.push(part);
        }
        return new Algorithm(algParts1);
    }

    // String based algorithm building methods
    SimpleVerbatimAlgorithm(...sayThis) {
        // Returns an algorithm that says the sayThis Strings verbatim per think cycle
        return this.AlgBuilder(new APVerbatim(...sayThis));
    }

    StrContainsList(str1, items) {
        // Returns the 1st match between words in a string and values in a list.
        for (const temp of items) {
            if (str1.includes(temp)) {
                return temp;
            }
        }
        return "";
    }

    SkillNotes(param) {
        return "notes unknown";
    }
}

class DiHelloWorld extends Skill {
    // hello world skill for testing purposes
    constructor() {
        super();
    }

    Input(ear, skin, eye) {
        switch (ear) {
            case "hello":
                super.SetVerbatimAlg(4, "hello world"); // 1->5 1 is the highest algorithm priority
                break;
        }
    }

    SkillNotes(param) {
        if (param === "notes") {
            return "plain hello world skill";
        } else if (param === "triggers") {
            return "say hello";
        }
        return "note unavailable";
    }
}

class Cerabellum {
    constructor() {
        this.fin = 0;
        this.at = 0;
        this.incrementAt = false;
        this.alg = null;
        this.ia = false; // isActive attribute
        this.emot = "";
    }

    AdvanceInAlg() {
        if (this.incrementAt) {
            this.incrementAt = false;
            this.at++;
            if (this.at === this.fin) {
                this.ia = false;
            }
        }
    }

    GetAt() {
        return this.at;
    }

    GetEmot() {
        return this.emot;
    }

    SetAlgorithm(algorithm) {
        if (!this.IsActive() && algorithm.GetAlgParts().length !== 0) {
            this.alg = algorithm;
            this.at = 0;
            this.fin = algorithm.GetSize();
            this.ia = true;
            this.emot = this.alg.GetAlgParts()[this.at].MyName(); // Updated line
            return false;
        }
        return true;
    }

    IsActive() {
        return this.ia;
    }

    Act(ear, skin, eye) {
        let axnStr = "";
        if (!this.IsActive()) {
            return axnStr;
        }
        if (this.at < this.fin) {
            axnStr = this.alg.GetAlgParts()[this.at].Action(ear, skin, eye);
            this.emot = this.alg.GetAlgParts()[this.at].MyName();
            if (this.alg.GetAlgParts()[this.at].Completed()) {
                this.incrementAt = true;
            }
        }
        return axnStr;
    }

    DeActivation() {
        this.ia = this.IsActive() && !this.alg.GetAlgParts()[this.at].algKillSwitch;
    }
}

class Fusion {
    constructor() {
        this.emot = "";
        this.result = "";
        this.ceraArr = Array.from({ length: 5 }, () => new Cerabellum());
    }

    GetEmot() {
        return this.emot;
    }

    LoadAlgs(neuron) {
        for (let i = 1; i <= 5; i++) {
            if (!this.ceraArr[i - 1].IsActive()) {
                const temp = neuron.GetAlg(i);
                if (temp !== null) {
                    this.ceraArr[i - 1].SetAlgorithm(temp);
                }
            }
        }
    }

    RunAlgs(ear, skin, eye) {
        this.result = "";
        for (let i = 0; i < 5; i++) {
            if (!this.ceraArr[i].IsActive()) {
                continue;
            }
            this.result = this.ceraArr[i].Act(ear, skin, eye);
            this.ceraArr[i].AdvanceInAlg();
            this.emot = this.ceraArr[i].GetEmot();
            this.ceraArr[i].DeActivation(); // Deactivation if Mutatable.algKillSwitch = true
            return this.result;
        }
        this.emot = "";
        return this.result;
    }
}

class Chobits {
    constructor() {
        this.dClasses = [];
        this.fusion = new Fusion();
        this.noiron = new Neuron();
        this.kokoro = new Kokoro(new AbsDictionaryDB()); // consciousness
        this.isThinking = false;
        this.awareSkills = [];
    }

    SetDataBase(absDictionaryDB) {
        this.kokoro = new Kokoro(absDictionaryDB);
    }

    AddSkill(skill) {
        // add a skill (builder design patterned func)
        if (this.isThinking) {
            return this;
        }
        skill.SetKokoro(this.kokoro);
        this.dClasses.push(skill);
        return this;
    }

    AddSkillAware(skill) {
        // add a skill with Chobit Object in their constructor
        skill.SetKokoro(this.kokoro);
        this.awareSkills.push(skill);
        return this;
    }

    ClearSkills() {
        // remove all skills
        if (this.isThinking) {
            return;
        }
        this.dClasses = [];
    }

    AddSkills(...skills) {
        if (this.isThinking) {
            return;
        }
        for (const skill of skills) {
            skill.SetKokoro(this.kokoro);
            this.dClasses.push(skill);
        }
    }

    RemoveSkill(skill) {
        if (this.isThinking) {
            return;
        }
        const index = this.dClasses.indexOf(skill);
        if (index > -1) {
            this.dClasses.splice(index, 1);
        }
    }

    ContainsSkill(skill) {
        return this.dClasses.includes(skill);
    }

    Think(ear, skin, eye) {
        this.isThinking = true;
        for (const dCls of this.dClasses) {
            this.InOut(dCls, ear, skin, eye);
        }
        this.isThinking = false;
        for (const dCls2 of this.awareSkills) {
            this.InOut(dCls2, ear, skin, eye);
        }
        this.fusion.LoadAlgs(this.noiron);
        return this.fusion.RunAlgs(ear, skin, eye);
    }

    GetSoulEmotion() {
        // get the last active AlgPart name
        // the AP is an action, and it also represents
        // an emotion
        return this.fusion.GetEmot();
    }

    InOut(dClass, ear, skin, eye) {
        dClass.Input(ear, skin, eye); // new
        dClass.Output(this.noiron);
    }

    GetKokoro() {
        // several chobits can use the same soul
        // this enables telepathic communications
        // between chobits in the same project
        return this.kokoro;
    }

    SetKokoro(kokoro) {
        // use this for telepathic communication between different chobits objects
        this.kokoro = kokoro;
    }

    GetFusion() {
        return this.fusion;
    }

    GetSkillList() {
        const result = [];
        for (const skill of this.dClasses) {
            result.push(skill.constructor.name);
        }
        return result;
    }
}

class Brain {
    constructor() {
        this.logicChobit = new Chobits();
        this.hardwareChobit = new Chobits();
        this.hardwareChobit.SetKokoro(this.logicChobit.GetKokoro());
        this.emotion = "";
        this.bodyInfo = "";
        this.logicChobitOutput = "";
    }

    get GetEmotion() {
        return this.emotion;
    }

    get GetBodyInfo() {
        return this.bodyInfo;
    }

    get GetLogicChobitOutput() {
        return this.logicChobitOutput;
    }

    DoIt(ear, skin, eye) {
        if (this.bodyInfo) {
            this.logicChobitOutput = this.logicChobit.Think(ear, this.bodyInfo, eye);
        } else {
            this.logicChobitOutput = this.logicChobit.Think(ear, skin, eye);
        }
        this.emotion = this.logicChobit.GetSoulEmotion();
        // Case: Hardware skill wishes to pass info to logical chobit
        this.bodyInfo = this.hardwareChobit.Think(this.logicChobitOutput, skin, eye);
    }

    AddLogicalSkill(skill) {
        this.logicChobit.AddSkill(skill);
    }

    AddHardwareSkill(skill) {
        this.hardwareChobit.AddSkill(skill);
    }
}

class DiPrinter extends Skill {
    // hello world skill for testing purposes
    constructor() {
        super();
    }

    Input(ear, skin, eye) {
        if (ear === "") {
            return;
        }
        console.log(ear);
    }
}


module.exports = { AbsDictionaryDB, Mutatable, APSay, APVerbatim, GrimoireMemento, Algorithm, Kokoro, Neuron, Skill, DiHelloWorld, Cerabellum, Fusion, Chobits, Brain, DiPrinter };
